### PR TITLE
stream info: add connection id

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: add connection ID to stream info, wire it through HCM and connection handler, add the property to access logs and request context.
Risk Level: low
Testing: unit tests
Docs Changes: yes
Release Notes: none

### CODE REVIEW

**Comment by asraa** [comment]
Looks like test coverage for the expr directory fell below, could you see if you could fix this? Thanks!

File: docs/root/intro/arch_overview/security/rbac_filter.rst [hunk 1]
```diff
@@ -99,6 +99,7 @@ The following attributes are exposed to the language runtime:
    connection.dns_san_peer_certificate, string, The first DNS entry in the SAN field of the peer certificate in the downstream TLS connection
    connection.uri_san_local_certificate, string, The first URI entry in the SAN field of the local certificate in the downstream TLS connection
    connection.uri_san_peer_certificate, string, The first URI entry in the SAN field of the peer certificate in the downstream TLS connection
+   connection.id, uint, Downstream connection ID
```
Comment (asraa) [inline docs/root/intro/arch_overview/security/rbac_filter.rst:102]: nit: is it accurate that it is 0 if not set? (and in the access log doc)

File: test/extensions/filters/common/expr/context_test.cc [hunk 2]
```diff
@@ -357,6 +357,31 @@ TEST(Context, ResponseAttributes) {
...
+  }
+
+  {
+    auto value = connection[CelValue::CreateStringView(ID)];
+    EXPECT_FALSE(value.has_value());
+  }
+
+  {
+    auto value = upstream[CelValue::CreateStringView(Undefined)];
+    EXPECT_FALSE(value.has_value());
```
Comment (asraa) [inline test/extensions/filters/common/expr/context_test.cc:376]: why is this repeated?

File: test/extensions/filters/common/expr/context_test.cc [hunk 3]
```diff
@@ -205,7 +205,7 @@ TEST(Context, RequestFallbackAttributes) {
   Http::TestRequestHeaderMapImpl header_map{
       {":method", "POST"},
       {":scheme", "http"},
-      {":path", "/meow?yes=1"},
+      {":path", "/meow"},
```
Comment (asraa) [inline test/extensions/filters/common/expr/context_test.cc:208]: what changed about this test?

**Review by asraa — CHANGES_REQUESTED** [review]
Thank you! Just some very minor comments. Thanks for adding the rbac test.

File: test/extensions/filters/common/expr/context_test.cc (hunk 2 above)
Comment (kyessenov) [reply test/extensions/filters/common/expr/context_test.cc:376]: It uses `upstream` instead of `connection` here.

File: test/extensions/filters/common/expr/context_test.cc (hunk 3 above)
Comment (kyessenov) [reply test/extensions/filters/common/expr/context_test.cc:208]: There is an edge case of missing `?` that gets covered.

File: docs/root/intro/arch_overview/security/rbac_filter.rst (hunk 1 above)
Comment (kyessenov) [reply docs/root/intro/arch_overview/security/rbac_filter.rst:102]: No, it becomes "unset", which is an error. `has(connection.id)` would return `false`. For access log, it looks it's reasonable to return 0 since I'm not sure how missing values are handled there.

Comment (asraa) [reply docs/root/intro/arch_overview/security/rbac_filter.rst:102]: Ah, gotcha. Yeah, for access logs the extractor needs a 0 value, like response codes setting 0 for unset values.

File: include/envoy/stream_info/stream_info.h [hunk 4]
```diff
@@ -577,6 +577,16 @@ class StreamInfo {
    * @return A shared pointer to the request ID utils for this stream
    */
   virtual Http::RequestIDExtensionSharedPtr getRequestIDExtension() const PURE;
+
+  /**
+   * @return Connection ID for the downstream connection.
```
Comment (asraa) [inline include/envoy/stream_info/stream_info.h:582]: Returns an absl::nullopt for unset?

**Review by asraa — COMMENTED** [review]
Thanks! (Sorry for the late review, I was out yesterday). LGTM besides questions about comments..

Quick pass to @envoyproxy/senior-maintainers for merge.

File: docs/root/configuration/observability/access_log/usage.rst [hunk 5]
```diff
@@ -356,6 +356,9 @@ The following command operators are supported:
 %DOWNSTREAM_LOCAL_ADDRESS_WITHOUT_PORT%
     Same as **%DOWNSTREAM_LOCAL_ADDRESS%** excluding port if the address is an IP address.
 
+%CONNECTION_ID%
+  ID of the downstream connection.
```
Comment (ggreenway) [inline docs/root/configuration/observability/access_log/usage.rst:360]: Please elaborate on what this value means and/or how it can be useful

**Review by ggreenway — CHANGES_REQUESTED** [review]

File: docs/root/configuration/observability/access_log/usage.rst (hunk 5 above)
Comment (kyessenov) [reply docs/root/configuration/observability/access_log/usage.rst:360]: We have two uses:
- cross-reference multiple access log sinks
- cross-reference timer-based log reporters

File: include/envoy/stream_info/stream_info.h (hunk 4 above)
Comment (kyessenov) [reply include/envoy/stream_info/stream_info.h:582]: Updated.

File: docs/root/configuration/observability/access_log/usage.rst (hunk 5 above)
Comment (kyessenov) [reply docs/root/configuration/observability/access_log/usage.rst:360]: Updated.

Comment (kyessenov) [reply docs/root/configuration/observability/access_log/usage.rst]: https://github.com/envoyproxy/envoy/issues/10936 will make it more useful in file access log context.

Comment (ggreenway) [reply docs/root/configuration/observability/access_log/usage.rst]: Ok, that makes sense in the context of the other PR. Can you document the ID's uniqueness? (Unique within an execution of envoy, can duplicate between multiple envoys, can duplicate after envoy restarts)

**Comment by kyessenov** [comment]
/retest
clang failed but not clear why (maybe killed?)

**Review by ggreenway — APPROVED** [review]

**Comment by ggreenway** [comment]
@kyessenov I don't know what happened to CI here. Can you merge master and push, mostly just to poke CI into running again?

File: docs/root/configuration/observability/access_log/usage.rst [hunk 6]
```diff
@@ -356,6 +356,13 @@ The following command operators are supported:
 %DOWNSTREAM_LOCAL_ADDRESS_WITHOUT_PORT%
     Same as **%DOWNSTREAM_LOCAL_ADDRESS%** excluding port if the address is an IP address.
 
+%CONNECTION_ID%
```
Comment (mattklein123) [inline docs/root/configuration/observability/access_log/usage.rst:359]: @kyessenov this needs a release note. Please do a follow up. Thanks!
